### PR TITLE
Custom gates updates

### PR DESCRIPTION
**Pull request by richrines1**
various updates/fixes to `custom_gates` (see also SupertechLabs/cirq-superstaq#245):
* adds `custom_gates.iXGate` to make `AQTiCCXGate` definition/signature/diagram more consistent
* modifies `ZZSwapGate.definition` so it has the same global phase as `ZZSwapGate.__array__`
* unrolls `ParallelGates` instances when they're a component of another `ParallelGates`
* corrects various type hints
* more robust tests

### CODE REVIEW

File: qiskit_superstaq/custom_gates.py [hunk 1]
```diff
@@ -118,11 +119,11 @@ def _define(self) -> None:
         qc = qiskit.QuantumCircuit(2, name="zzswap")
         qc.cx(0, 1)
         qc.cx(1, 0)
-        qc.rz(self.params[0], 1)
+        qc.p(self.params[0], 1)
         qc.cx(0, 1)
         self.definition = qc
 
-    def __array__(self, dtype: Type = None) -> np.ndarray:
+    def __array__(self, dtype: Optional[type] = None) -> np.ndarray:
```
Comment (vtomole) [inline qiskit_superstaq/custom_gates.py:126]: Are we able to turn on a mypy setting that would always flag a case like this?

File: qiskit_superstaq/custom_gates.py [hunk 2]
```diff
@@ -153,14 +154,21 @@ def __init__(self, *component_gates: qiskit.circuit.Gate, label: Optional[str] =
             component_gates: Gate(s) to be collected into single gate
             label: an optional label for the constructed Gate
         """
-        if not all(isinstance(gate, qiskit.circuit.Gate) for gate in component_gates):
-            raise ValueError("Component gates must be instances of qiskit.circuit.Gate")
+        self.component_gates: List[qiskit.circuit.Gate] = []
```
Comment (vtomole) [inline qiskit_superstaq/custom_gates.py:157]: If i recall, we decided to make this a tuple in css

Comment (richrines1) [reply qiskit_superstaq/custom_gates.py]: yep good catch, fixed.

File: qiskit_superstaq/custom_gates.py (hunk 1 above)
Comment (richrines1) [reply qiskit_superstaq/custom_gates.py:126]: we can set `no_implicit_optional = True` in setup.cfg to require the use of `Optional`. it requires a number of changes elsewhere, so filed . for a subsequent PR

File: qiskit_superstaq/custom_gates.py [hunk 3]
```diff
@@ -244,50 +298,62 @@ def _define(self) -> None:
...
-    return None
+
+    if gate.definition and gate.definition.name in _custom_gate_resolvers:
+        new_gate = _custom_gate_resolvers[gate.definition.name](*gate.params)
+    elif gate.name in _custom_gate_resolvers:
+        new_gate = _custom_gate_resolvers[gate.name](*gate.params)
+    else:
+        return None
+
+    new_gate.label = gate.label
```
Comment (singular-value) [inline qiskit_superstaq/custom_gates.py:358]: Nice

Since this wasn't in the previous code, can you add a test (a regression test I guess) for it?

Comment (richrines1) [reply qiskit_superstaq/custom_gates.py:358]: done

**Comment by singular-value** [comment]
lgtm, leaving to @vtomole

**Review by vtomole — APPROVED** [review]
LGTM

**Pull request merged**